### PR TITLE
impl(pubsub): add thread id attribute to publish span

### PR DESCRIPTION
**Pull request by alevenberg**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13151)

### CODE REVIEW

File: google/cloud/pubsub/internal/tracing_message_batch.cc [hunk 1]
```diff
@@ -43,6 +44,11 @@ using Attributes =
 using Links =
     std::vector<std::pair<opentelemetry::trace::SpanContext, Attributes>>;
 
+int64_t GetCurrentThreadId() {
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/tracing_message_batch.cc:47]: Refactor and use this?

https://github.com/googleapis/google-cloud-cpp/blob/45b5c659b146729b1bb85ad2f64d2218265a7b8c/google/cloud/storage/internal/async/writer_connection_tracing.cc#L33-L37

Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch.cc:47]: The otel recc says that the attribute should be an [int](https://opentelemetry.io/docs/specs/semconv/general/attributes/#general-thread-attributes). I think the other spot should be refactored. 

By refactor, do you mean make this a common function in opentelemetry.h?

Comment (coryan) [reply google/cloud/pubsub/internal/tracing_message_batch.cc:47]: I wonder how they expect to get an `int` out of [std::thread::id](https://en.cppreference.com/w/cpp/thread/thread/id).  There is no guaranteed conversion to any integral type.  Just formatting to `std::iostream`.

Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch.cc:47]: https://github.com/open-telemetry/semantic-conventions/issues/546

File: google/cloud/storage/internal/async/reader_connection_tracing.cc [hunk 2]
```diff
@@ -31,12 +31,6 @@ namespace {
 
 namespace sc = ::opentelemetry::trace::SemanticConventions;
 
-std::string CurrentThreadId() {
-  std::ostringstream os;
```
Comment (devbww) [inline google/cloud/storage/internal/async/reader_connection_tracing.cc:35]: Removes the need to include `<sstream>`.

File: google/cloud/storage/internal/async/writer_connection_tracing.cc [hunk 3]
```diff
@@ -30,12 +30,6 @@ namespace {
 
 namespace sc = ::opentelemetry::trace::SemanticConventions;
 
-std::string CurrentThreadId() {
-  std::ostringstream os;
```
Comment (devbww) [inline google/cloud/storage/internal/async/writer_connection_tracing.cc:34]: Removes the need to include `<sstream>`.

File: google/cloud/pubsub/internal/tracing_message_batch.cc [hunk 4]
```diff
@@ -24,6 +24,7 @@
 #include "opentelemetry/trace/span.h"
 #include <algorithm>
 #include <string>
+#include <thread>
```
Comment (devbww) [inline google/cloud/pubsub/internal/tracing_message_batch.cc:27]: Remnant?

File: google/cloud/internal/opentelemetry.h [hunk 5]
```diff
@@ -216,6 +216,9 @@ std::string ToString(opentelemetry::trace::TraceId const& trace_id);
 
 std::string ToString(opentelemetry::trace::SpanId const& span_id);
 
+/// Gets the current thread id.
+std::string CurrentThreadId();
```
Comment (devbww) [inline google/cloud/internal/opentelemetry.h:220]: `#include <string>`

File: google/cloud/internal/opentelemetry.cc [hunk 6]
```diff
@@ -136,6 +137,12 @@ std::string ToString(opentelemetry::trace::SpanId const& span_id) {
   return std::string(span_id_array, kSize);
 }
 
+std::string CurrentThreadId() {
+  std::ostringstream os;
+  os << std::this_thread::get_id();
```
Comment (devbww) [inline google/cloud/internal/opentelemetry.cc:142]: `#include <thread>`

**Comment by alevenberg** [comment]
Resolved comments I addressed

File: google/cloud/pubsub/internal/tracing_message_batch_test.cc [hunk 7]
```diff
@@ -147,6 +147,29 @@ TEST(TracingMessageBatch, Flush) {
                                  "messaging.pubsub.message.link", 0)))))));
 }
 
+TEST(TracingMessageBatch, PublishSpanHasThreadIdAttribute) {
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/tracing_message_batch_test.cc:150]: There is something funny about verifying that the "publish" span has the `sc::kCodeFunction` attribute in every test, but having a dedicated test for the `sc::kThreadId` attribute.

**Review by dbolduc — APPROVED** [review]

**Comment by alevenberg** [comment]
Fixed the operation attribute test too

File: google/cloud/pubsub/internal/tracing_message_batch_test.cc (hunk 7 above)
Comment (dbolduc) [reply google/cloud/pubsub/internal/tracing_message_batch_test.cc:148]: Consider combining these into one test: `PublishSpanAttributes`.

**Review by dbolduc — APPROVED** [review]

**Pull request merged**